test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterDerivedTypeConstraintTests.cs

```diff
@@ -731,5 +731,113 @@ private static void SetDerivedTypeAnnotation(EdmModel model, IEdmVocabularyAnnot
             valueAnnotationOnProperty.SetSerializationLocation(model, EdmVocabularyAnnotationSerializationLocation.Inline);
             model.AddVocabularyAnnotation(valueAnnotationOnProperty);
         }
+
+        #region Serialize invalid type resource
+        [Fact]
+        public void WritingResourceWithTypeInHeritanceTreeWithResourceSetTypeSuccessed()
+        {
+            // Arrange
+            var model = GetGraphModel();
+            var directoryObjects = model.EntityContainer.FindEntitySet("directoryObjects");
+            var directoryObject = model.SchemaElements.OfType<IEdmEntityType>().FirstOrDefault(c => c.Name == "directoryObject");
+
+            ODataResource group = new ODataResource
+            {
+                TypeName = "microsoft.graph.group",
+                Properties = new[] { new ODataProperty { Name = "id", Value = "abc" }, new ODataProperty { Name = "displayName", Value = "xyz" } }
+            };
+
+            Action<ODataMessageWriter> writeAction = (messageWriter) =>
+            {
+                ODataWriter writer = messageWriter.CreateODataResourceSetWriter(directoryObjects, directoryObject); // entityset
+                writer.WriteStart(new ODataResourceSet());
+                writer.WriteStart(group);
+                writer.WriteEnd();
+                writer.WriteEnd();
+            };
+
+            // Act
+            string actual = GetWriterOutput(model, writeAction);
+
+            // Assert
+            Assert.Equal("{\"@odata.context\":\"http://example.com/$metadata#directoryObjects\",\"value\":[{\"@odata.type\":\"#microsoft.graph.group\",\"id\":\"abc\",\"displayName\":\"xyz\"}]}",
+                actual);
+        }
+
+        [Fact]
+        public void WritingResourceWithTypeNotInHeritanceTreeWithResourceSetTypeThrows()
+        {
+            // Arrange
+            var model = GetGraphModel();
+            var directoryObjects = model.EntityContainer.FindEntitySet("directoryObjects");
+            var directoryObject = model.SchemaElements.OfType<IEdmEntityType>().FirstOrDefault(c => c.Name == "directoryObject");
+
+            ODataResource contact = new ODataResource
+            {
+                TypeName = "microsoft.graph.contact",
+                Properties = new[] { new ODataProperty { Name = "id", Value = "abc" }, new ODataProperty { Name = "displayName", Value = "xyz" } } /*open type*/
+            };
+
+            Action<ODataMessageWriter> writeAction = (messageWriter) =>
+            {
+                ODataWriter writer = messageWriter.CreateODataResourceSetWriter(directoryObjects, directoryObject); // entityset
+                writer.WriteStart(new ODataResourceSet());
+                    writer.WriteStart(contact);

```

Comment:
Is this intentionally indented in this way?

Reply:
Yes